### PR TITLE
lexer: add XML-style comments and fix line counter

### PR DESCRIPTION
**Pull request by step-**
This commit adds XML-style comments as a lexical feature that script
developers can use to easily document their code. It also fixes two
long-standing bugs: the line counter that references the location of
a syntax error did not count multi-line values; markup `<edit ` was
incorrectly reported as an `<entry>` token (with no ill consequences).

Prior to this commit the gtkdialog markup did not provide a native way
to enter comments. Several work-arounds were in use: side-loading
comments in a separate file; writing shell comments at the top/bottom/
middle of the shell script file; embedding shell comments in gtkdialog
markup using various ad-hoc (and fragile) lexical hacks; using a shell
function to create an envelope for comments; forking an `<action>` that
runs nothing but shell comments. These methods are still possible but
are discouraged for new scripts.

**BREAKING CHANGE**

Gtkdialog versions <= 0.8.4 can't parse XML-style comments and error out
so this feature is not backward-compatible.
See the new script `examples/miscellaneous/comments.sh` for ways you
can use - and not use - comments in your new scripts.

### CODE REVIEW

**Review by 01micko — COMMENTED** [review]
I think this is a _very_ useful feature. Certainly worthy of the `0.8.5` release tag.

I'll approve but wait for @dimkr 's analysis.

**Review by 01micko — APPROVED** [review]

**Comment by dimkr** [comment]
You're so productive, @step-! I can't find the time to review everything 😄

**Comment by 01micko** [comment]
> You're so productive, @step-! I can't find the time to review everything smile

Well, I _finally_ found some time to invite @step- to the organization, so maybe when I do a PR he can review it!

**Comment by step-** [comment]
> Well, I _finally_ found some time to invite @step- to the organization, so maybe when I do a PR he can review it!

LOL

**Comment by 01micko** [comment]
You now have write access to `gtkdialog` @step- .. you can even merge your own PRs :rofl:

**Comment by step-** [comment]
> You now have write access to `gtkdialog` @step- .. you can even merge your own PRs rofl

The ultimate perversion?  :)

**Comment by step-** [comment]
> You now have write access to `gtkdialog` @step- .. you can even merge your own PRs rofl

@dimkr, @01micko I could merge this PR if @dimkr chooses to pass on his review but first I need to emphasize that using XML-style comments in a script makes the script incompatible with gtkdialog version 0.8.4.  While features introduced in recent PRs degraded gracefully, this one doesn't.  I don't see this as an issue but you might think otherwise.

**Comment by dimkr** [comment]
> @dimkr, @01micko I could merge this PR if @dimkr chooses to pass on his review but first I need to emphasize that using XML-style comments in a script makes the script incompatible with gtkdialog version 0.8.4.

This means we'll have to avoid using this feature, at least until all woof-CE users start using the gtkdialog petbuild and get migrated to 0.8.5.

**Comment by step-** [comment]
> This means we'll have to avoid using this feature, at least until all woof-CE users start using the gtkdialog petbuild and get migrated to 0.8.5. As long as it doesn't break anything that doesn't have comments inside the XML, I think I'm fine with this change, because it's not a breaking change for existing stuff.

Correct.

**Comment by 01micko** [comment]
I think we merge this and as responsible Puppy family developers (including *dogs and other offshoots) abstain from xml comments for a grace period til say 2024?

**Comment by 01micko** [comment]
> That said, let's leave a script developer choose whether to use these new features knowing that they aren't backward compatible.

Sounds good, merging.

**Comment by dimkr** [comment]
> > That said, let's leave a script developer choose whether to use these new features knowing that they aren't backward compatible.
> 
> Sounds good, merging.

Maybe we should write a linter that makes sure a gtkdialog-based script is backwards compatible with 0.8.4, and run it in woof-CE's CI. It's easy to identify scripts that make use of gtkdialog (`fgrep '<window'` ?), and it should be equally easy to make sure they don't have XML comments.

**Comment by step-** [comment]
> Maybe we should write a linter that makes sure a gtkdialog-based script is backwards compatible with 0.8.4, and run it in woof-CE's CI. It's easy to identify scripts that make use of gtkdialog (`fgrep '<window'` ?), and it should be equally easy to make sure they don't have XML comments.

I was toying with the same idea just yesterday though I didn't go as far as considering integrating with woof-CE's CI.  I was thinking more of a pre-flighting tool for gtkdialog scripts. Removing comments is simple enough.  Transforming input redirection (#127) may be possible - I'd have to think about it more.  Thanks for your idea.

**Pull request merged**
